### PR TITLE
fix: use match to identify test configuration

### LINKED ISSUE #390: Please check contains "test" for `skipTestGroups`, not starts with
Currently, the `skipTestGroups` only checks for configurations starting with `androidTest` or `test`.
But this does not match things like `integTest...`, `systemTest...`, `functionalTest...`, ...
I think for the default recognition it would be more appropriate to check for `toLowerCase().contains("test")` instead.

To do something like that manually, you need some discouraged work-arounds like in the comments of #22 that reach into the models of all other projects to get all configuration names and then be able to filter on them, or list them all in the root project and hope non get added in subprojects in the future.

### PR DESCRIPTION
**Pull request by jeremylong**
resolves #390

### CODE REVIEW

File: src/main/groovy/org/owasp/dependencycheck/gradle/tasks/AbstractAnalyze.groovy [hunk 1]
```diff
@@ -377,18 +377,15 @@ abstract class AbstractAnalyze extends ConfiguredTask {
...
-     *     <li>the configuration name starts with 'test'</li>
-     *     <li>the configuration name starts with 'androidTest'</li>
+     *     <li>the name of the configuration or any of its parent configurations matches /^(.*[a-z0-9_]T|_?t)est([A-Z0-9_].*)?$/</li>
      * </ul>
+     * The intent of the regular expression is to match `test` in a camel case or snake case configuration name.
      */
     @groovy.transform.CompileStatic
     static boolean isTestConfigurationCheck(Configuration configuration) {
-        boolean isTestConfiguration = configuration.name.startsWith("test") || configuration.name.startsWith("androidTest")
+        boolean isTestConfiguration = configuration.name.matches('^(.*[a-z0-9_]T|_?t)est([A-Z0-9_].*)?$')
```
Comment (jeremylong) [inline src/main/groovy/org/owasp/dependencycheck/gradle/tasks/AbstractAnalyze.groovy:386]: be more groovy
```suggestion
        boolean isTestConfiguration = configuration.name =~ /((^|[a-z0-9_])T|(^|_)t)est([A-Z0-9_]|$)/
```

File: src/main/groovy/org/owasp/dependencycheck/gradle/tasks/AbstractAnalyze.groovy [hunk 2]
```diff
@@ -377,18 +377,15 @@ abstract class AbstractAnalyze extends ConfiguredTask {
...
         configuration.hierarchy.each {
-            isTestConfiguration |= (it.name == "testCompile" || it.name == "androidTestCompile" || it.name == "testImplementation")
+            isTestConfiguration |= it.name.matches('^(.*[a-z0-9_]T|_?t)est([A-Z0-9_].*)?$')
```
Comment (jeremylong) [inline src/main/groovy/org/owasp/dependencycheck/gradle/tasks/AbstractAnalyze.groovy:388]: be more groovy
```suggestion
            isTestConfiguration |= it.name =~ /((^|[a-z0-9_])T|(^|_)t)est([A-Z0-9_]|$)/
```

File: src/main/groovy/org/owasp/dependencycheck/gradle/tasks/AbstractAnalyze.groovy [hunk 3]
```diff
@@ -377,18 +377,15 @@ abstract class AbstractAnalyze extends ConfiguredTask {
...
      * </ul>
+     * The intent of the regular expression is to match `test` in a camel case or snake case configuration name.
      */
     @groovy.transform.CompileStatic
     static boolean isTestConfigurationCheck(Configuration configuration) {
-        boolean isTestConfiguration = configuration.name.startsWith("test") || configuration.name.startsWith("androidTest")
+        boolean isTestConfiguration = configuration.name =~ /((^|[a-z0-9_])T|(^|_)t)est([A-Z0-9_]|$)/
         configuration.hierarchy.each {
-            isTestConfiguration |= (it.name == "testCompile" || it.name == "androidTestCompile" || it.name == "testImplementation")
+            isTestConfiguration |= it.name =~ /((^|[a-z0-9_])T|(^|_)t)est([A-Z0-9_]|$)/
```
Comment (jeremylong) [inline src/main/groovy/org/owasp/dependencycheck/gradle/tasks/AbstractAnalyze.groovy:388]:
```suggestion
            isTestConfiguration |= it.name ==~ /((^|[a-z0-9_])T|(^|_)t)est([A-Z0-9_]|$)/
```

File: src/main/groovy/org/owasp/dependencycheck/gradle/tasks/AbstractAnalyze.groovy [hunk 4]
```diff
@@ -377,18 +377,15 @@ abstract class AbstractAnalyze extends ConfiguredTask {
...
-     *     <li>the configuration name starts with 'test'</li>
-     *     <li>the configuration name starts with 'androidTest'</li>
+     *     <li>the name of the configuration or any of its parent configurations matches /^(.*[a-z0-9_]T|_?t)est([A-Z0-9_].*)?$/</li>
      * </ul>
+     * The intent of the regular expression is to match `test` in a camel case or snake case configuration name.
      */
     @groovy.transform.CompileStatic
     static boolean isTestConfigurationCheck(Configuration configuration) {
-        boolean isTestConfiguration = configuration.name.startsWith("test") || configuration.name.startsWith("androidTest")
+        boolean isTestConfiguration = configuration.name =~ /((^|[a-z0-9_])T|(^|_)t)est([A-Z0-9_]|$)/
```
Comment (jeremylong) [inline src/main/groovy/org/owasp/dependencycheck/gradle/tasks/AbstractAnalyze.groovy:386]:
```suggestion
        boolean isTestConfiguration = configuration.name ==~ /((^|[a-z0-9_])T|(^|_)t)est([A-Z0-9_]|$)/
```

File: src/main/groovy/org/owasp/dependencycheck/gradle/tasks/AbstractAnalyze.groovy [hunk 5]
```diff
@@ -377,18 +377,15 @@ abstract class AbstractAnalyze extends ConfiguredTask {
...
-     *     <li>the configuration name starts with 'test'</li>
-     *     <li>the configuration name starts with 'androidTest'</li>
+     *     <li>the name of the configuration or any of its parent configurations matches /^(.*[a-z0-9_]T|_?t)est([A-Z0-9_].*)?$/</li>
      * </ul>
+     * The intent of the regular expression is to match `test` in a camel case or snake case configuration name.
      */
     @groovy.transform.CompileStatic
     static boolean isTestConfigurationCheck(Configuration configuration) {
-        boolean isTestConfiguration = configuration.name.startsWith("test") || configuration.name.startsWith("androidTest")
+        boolean isTestConfiguration = configuration.name ==~ /((^|[a-z0-9_])T|(^|_)t)est([A-Z0-9_]|$)/
```
Comment (jeremylong) [inline src/main/groovy/org/owasp/dependencycheck/gradle/tasks/AbstractAnalyze.groovy:386]:
```suggestion
        boolean isTestConfiguration = configuration.name =~ /((^|[a-z0-9_])T|(^|_)t)est([A-Z0-9_]|$)/
```

File: src/main/groovy/org/owasp/dependencycheck/gradle/tasks/AbstractAnalyze.groovy [hunk 6]
```diff
@@ -377,18 +377,15 @@ abstract class AbstractAnalyze extends ConfiguredTask {
      * Checks whether a configuration is considered to be a test configuration in order to skip it.
      * A configuration is considered a test configuration if and only if any of the following conditions holds:
      * <ul>
-     *     <li>the name of the configuration or any of its parent configurations equals 'testCompile'</li>
-     *     <li>the name of the configuration or any of its parent configurations equals 'testImplementation'</li>
-     *     <li>the name of the configuration or any of its parent configurations equals 'androidTestCompile'</li>
-     *     <li>the configuration name starts with 'test'</li>
-     *     <li>the configuration name starts with 'androidTest'</li>
+     *     <li>the name of the configuration or any of its parent configurations matches /^(.*[a-z0-9_]T|_?t)est([A-Z0-9_].*)?$/</li>
```
Comment (Vampire) [inline src/main/groovy/org/owasp/dependencycheck/gradle/tasks/AbstractAnalyze.groovy:380]: Don't forget changing to the new regex here. :-)

Comment (Vampire) [reply src/main/groovy/org/owasp/dependencycheck/gradle/tasks/AbstractAnalyze.groovy:380]: Also "matches" is no longer correct

Comment (jeremylong) [reply src/main/groovy/org/owasp/dependencycheck/gradle/tasks/AbstractAnalyze.groovy:380]:
```suggestion
     *     <li>the name of the configuration or any of its parent configurations matches /((^|[a-z0-9_])T|(^|_)t)est([A-Z0-9_]|$)/</li>
```

**Pull request merged**